order.qmd

```diff
@@ -0,0 +1,262 @@
+```
+
+however in more complex models such an approach may become unwieldy, and further changes to updates for `X` and `Y` must be replicated in the update to `Z`.
+
+A safer approach is to introduce intermediate variables representing the updated values of `X` and `Y`
+
+```{r}
+m <- odin({
+  update(X) <- X_new
+  update(Y) <- Y_new
+  update(Z) <- X_new + Y_new
+  
+  dX <- Normal(0, 1)
+  X_new <- X + dX
+  dY <- Normal(0, 1)
+  Y_new <- Y + dY
+  
+  initial(X) <- 0
+  initial(Y) <- 0
+  initial(Z) <- 0
+})
+
+```
+
+which means that subsequent changes to how `X` updates are automatically factored into how `Z` updates. We can see this version no longer results in a lag:
+
+```{r}
+sys <- dust_system_create(m, list(), seed = 1)
+dust_system_set_state_initial(sys)
+t <- seq(0, 5)
+y <- dust_system_simulate(sys, t)
+y <- dust_unpack_state(sys, y)
+
+y$X + y$Y
+
+y$Z
+```
+:::
+
+
+### Update time to t0 + dt
+
+The series of updates ends with `time` being increased by `dt`.
+
+
+## Comparison to data {#sec-order-comparison}
+
+The comparison to data follows on from the above series of updates, and as such you may find it is good practice to structure your `odin` code such that all equations relating to the comparison to data appear in a block at the end.
+
+Equations relating to comparison to data are evaluated after the update of time to `t0 + dt` and will only be evaluated if you have any data at `time = t0 + dt`.
+
+### Read from variables
+
+All variables will be read at their current values, which are the values corresponding to `time = t0 + dt`. Again, in our example the variables are `S`, `I`, `R` and `incidence`, although only `incidence` is used in the comparison to data
+
+### Look up interpolation
+
+Any uses of `interpolate` that are used in the comparison to data will be evaluated using `time = t0 + dt` (not using `time = t0` as was used earlier in the "update" step). Compilation of `odin` code will detect which interpolations are used in the comparison to data.
+
+In our example we 
+```r
+phi <- interpolate(phi_time, phi_value, "constant")
+```
+where `phi` might represent a time-varying ascertainment rate of cases (which could be affected by e.g. whether schools are open, or day of the week effects).
+
+### Evaluate assignments
+
+As with interpolation, compliation of `odin` code will detect which assignments are used only in the comparison to data. For instance in our example this would be
+
+```r
+exp_noise <- Exponential(1e6)
+```
+
+Any use of variables in these equations on the right-hand side will be evaluated using their values at `time = t0 + dt`.
+
+
+### Compare to data
+
+Finally we evaluate the density given by any relationship equations. In our example this is
+
+```r
+cases ~ Poisson(phi * incidence + exp_noise)
+```
+
+As with assigments in the comparison to data step, any use of variables in these equations on the right-hand side will be evaluated using their values at `time = t0 + dt`.
```

Comment:
Perhaps comment that no actual assignment is done, but the density accumulates over all non-NA data entries